package/MDAnalysis/coordinates/chemfiles.py

```diff
@@ -381,3 +382,63 @@ def _topology_to_chemfiles(self, obj, n_atoms):
             topology.add_bond(bond.atoms[0].ix, bond.atoms[1].ix)
 
         return topology
+
+
+if HAS_CHEMFILES:
+    class ChemfilesPicklable(chemfiles.Trajectory):
+        """Chemfiles file object (read-only) that can be pickled.
+
+        This class provides a file-like object (as returned by
+        :class:`chemfiles.Trajectory`) that,
+        unlike standard Python file objects,
+        can be pickled. Only read mode is supported.
+
+        When the file is pickled, path, mode, and format of the file handle
+        are saved. On unpickling, the file is opened by path with mode,
+        and saved format.
+        This means that for a successful unpickle, the original file still has
+        to be accessible with its filename.
+
+        Note
+        ----
+        Upon pickling, the current frame is reset. `universe.trajectory[i]` has
+        to be used to return to its original frame.
+
+        Parameters
+        ----------
+        filename : str
+            a filename given a text or byte string.
+        mode : 'r' , optional
+            can only be 'r' for pickling.
+        format : '', optional
+            guessed from the file extension if empty.
+
+        Example
+        -------
+        ::
+
+            f = ChemfilesPicklable(XYZ, 'r', '')
+            print(f.read())
+            f.close()
+
+        can also be used as context manager::
+
+            with ChemfilesPicklable(XYZ) as f:
+                print(f.read())
+
+        See Also
+        ---------
+        :class:`MDAnalysis.lib.picklable_file_io.FileIOPicklable`
+        :class:`MDAnalysis.lib.picklable_file_io.BufferIOPicklable`
+        :class:`MDAnalysis.lib.picklable_file_io.TextIOPicklable`
+        :class:`MDAnalysis.lib.picklable_file_io.GzipPicklable`
+        :class:`MDAnalysis.lib.picklable_file_io.BZ2Picklable`
+
+
+        .. versionadded:: 2.0.0
+        """
+        def __getstate__(self):
+            return self.path, self._Trajectory__mode, self._Trajectory__format
```

Comment:
You may also want to save the current step being read (self.step) to reset the file cursor. Although I am not sure how this fits within the largest framework here, so it might be fine.

Other pieces of state that could be set & restored are a custom cell and custom topology; but these are not exposed in MDA so this should be advanced usage anyway

Reply:
I may have misinterpreted something, but `chemfiles.Trajectory.step` does not exist. I checked the documents, and it mentioned 
>property nsteps
Get the current number of steps in this Trajectory.

But it seems that nsteps returns the total number of frames?

Reply:
Okay, I can get `step` from `Frame`. I am wondering if there's a way to get the current frame? Sorry I am not quite familiar with `chemfiles`

Reply:
No, there is no way to get the last read step directly from the trajectory. You could perhaps directly define __getstate__ on ChemfilesReader instead; and get the step from there as `self._step`

Reply:
Thanks! That's fine, after pickling, the current step for the reader will be set to trajectory[previous_ts] (this `__setstate__` is defined directly to `ReaderBase`).

Reply:
@Luthaf would it be possible to push these changes up to Chemfiles directly?  I.e. make `chemfiles.Trajectory` picklable?

Reply:
I would not be against it in general, although the approach taken here of serializing only the file name & hoping it stays accessible feels a bit fragile. 

My understanding is that this PR main goal is enabling Python multiprocessing, am I right? The approach should work fine for this use case, but not for general Python pickling. I am a bit hesitant of making the `chemfiles.Trajectory` pickable with this huge caveat, since it might break in strange ways for users who expect pickle to "just work"